src/main/java/org/zalando/nakadi/service/subscription/SubscriptionTimeLagService.java

```diff
@@ -0,0 +1,130 @@
+package org.zalando.nakadi.service.subscription;
+
+import com.google.common.collect.ImmutableList;
+import org.echocat.jomon.runtime.concurrent.RetryForSpecifiedTimeStrategy;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.zalando.nakadi.domain.ConsumedEvent;
+import org.zalando.nakadi.domain.EventTypePartition;
+import org.zalando.nakadi.domain.NakadiCursor;
+import org.zalando.nakadi.domain.PartitionEndStatistics;
+import org.zalando.nakadi.exceptions.ErrorGettingCursorTimeLagException;
+import org.zalando.nakadi.exceptions.InvalidCursorException;
+import org.zalando.nakadi.exceptions.NakadiException;
+import org.zalando.nakadi.exceptions.runtime.InconsistentStateException;
+import org.zalando.nakadi.repository.EventConsumer;
+import org.zalando.nakadi.service.NakadiCursorComparator;
+import org.zalando.nakadi.service.timeline.TimelineService;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import static org.echocat.jomon.runtime.concurrent.Retryer.executeWithRetry;
+
+@Component
+public class SubscriptionTimeLagService {
+
+    private static final int EVENT_FETCH_WAIT_TIME_MS = 1000;
+    private static final int COMPLETE_TIMEOUT_MS = 30000;
+    private static final int LAG_CALCULATION_PARALLELISM = 10;
+
+    private final TimelineService timelineService;
+    private final NakadiCursorComparator cursorComparator;
+
+    @Autowired
+    public SubscriptionTimeLagService(final TimelineService timelineService,
+                                      final NakadiCursorComparator cursorComparator) {
+        this.timelineService = timelineService;
+        this.cursorComparator = cursorComparator;
+    }
+
+    public Map<EventTypePartition, Duration> getTimeLags(final Collection<NakadiCursor> committedPositions,
+                                                         final List<PartitionEndStatistics> endPositions)
+            throws ErrorGettingCursorTimeLagException, InconsistentStateException {
+
+        final ExecutorService executor = Executors.newFixedThreadPool(LAG_CALCULATION_PARALLELISM);
```

Comment:
Does it make sense to constantly create threads in packs of 10? 
Why not to use single CachedThreadPool for this service?

Reply:
Fixed thread pool will create 10 Threads only if they are really needed (subscription has >= 10 partitions). It will anyway add threads 1 by 1 until reaches 10. The only problem it has is that it will never decrease thread num after that, but for us it's not a problem because the lifetime of this threadpool is very short.
 
And I don't see how to limit CachedThreadPool size.

Reply:
@v-stepanov 
Sometimes I feel myself as alien. I mean why not to use single cached thread pool for instance, not for request 
```
new ThreadPoolExecutor(
    0, // Minimum number of threads
    1000, // Maximum number of threads
    60L, // Idle timeout to kill threads after
    TimeUnit.SECONDS, // Idle timeout unit
    new LinkedBlockingQueue<>()) // Queue, to which tasks are scheduled
```
Let's suppose that I am doing 2000 requests targeting 8000 partitions. You will create definitely more than 2000 threads, at most - 8000 (will be pretty close to max), and after starting - will destroy it. With single cached thread pool you will create up to max parallel requests* avg partition count = 20 *4 = 80 (it is really expected value, in total). 
We can calculate difference and think about time spent for threads creation/termination. 

I don't really see the reason of single request limit to 10 threads. You are just saying that overall I have possibility of infinite thread creation (across different requests), but I will not try to accomplish fast this huge request, because I do not want to. 
This 10 limit will not save instance from 1000 concurrent requests, but will definitely make some requests slower. 
Also, having limit of threads dedicated for this operation will guarantee that nakadi will not be overloaded with this amount of requests.

Reply:
@antban I reworked it.
Now there is a common thread pool. But I anyway implemented the throttling per request (one request will not occupy more than 20 threads).
I hope this makes you satisfied.